doc/specs/#147 - Support different release channels [released, beta, alpha].md

```diff
@@ -0,0 +1,259 @@
+---
+author: Kaleb Luedtke - Trenly <trenlymc@gmail.com>
+created on: 2021-11-18
+last updated: 2021-11-20
+issue id: 147
+---
+
+# Release Channels
+
+For [#147](https://github.com/microsoft/winget-cli/issues/147)
+
+## Abstract
+
+Several packages have different channels for releasing different versions of their software. The Windows Package Manager should be able to support packages of the same Package Identifier where the versions come from different release channels. The Windows Package Manager should be also able to inform a user about which channels are available.
+
+## Inspiration
+
+Some other package managers have this capability. In the current state, a new package identifier must be created for each release channel which can cause conflicts when the versions are *not* side-by-side compatible.
+
+## Solution Design
+
+### Manifest Structure
+The Windows Package Manager manifest v1.1.0 schema has provided a key for declaring which channel an installer belongs to. However, this has the potential to make the installer manifests extremely complex when there are multiple release channels with multiple installers each.
+> The `Channel` key should be removed from the manifest v1.1.0 schema.
+
+Instead, the use of multiple installer manifest files will be used to define and manage the various channels available for a package in a way similar to locales. This will ensure that the client remains backwards compatible with old manifest versions, and will make the creation and maintenance of the various channels easier for contributors in the community repository. This will require a key be added to the  manifest schema. Channels should not be compatible with singleton manifests.
+> The `DefaultChannel` key should be added to the manifest v1.1.0 schema for version manifests
+
+An example of the manifest file structure, with installer channel manifests:
+```raw
+/manifests/g/Google/Chrome/1.0
+  | Google.Chrome.installer.beta.yaml
+  | Google.Chrome.installer.canary.yaml
+  | Google.Crome.installer.dev.yaml
+  | Google.Chrome.installer.Stable.yaml
+  | Google.Chrome.locale.en-US.yaml
+  | Google.Chrome.locale.nb-NO.yaml
+  | Google.Chrome.yaml
+/manifests/g/Google/Chrome/2.0
+  | Google.Chrome.installer.canary.yaml
+  | Google.Crome.installer.dev.yaml
+  | Google.Chrome.locale.en-US.yaml
+  | Google.Chrome.locale.nb-NO.yaml
+  | Google.Chrome.yaml
+```
+
+To maintain backwards compatability, `null` should be treated as a valid `DefaultChannel`. If the default channel is not specified, `<PackageIdentifier>.installer.yaml` would be treated as the default. Additional channels may still be added to the package, and all functionality would still be present.
```

Comment:
The code already expects that an empty value is the default channel.  Having a user defined default channel value will complicate things.  I don't see the value in allowing it even if it was easy either.

Reply:
> The original thought (pre multi-file manifests) was to have a structure like:
> 
> ```
> publisher\package\version
> publisher\package\non-default-channel\version
> ```
>  With Channel being in the installer schema, your design can work too. I think it might be a bit more confusing though, as it is harder to see what is in each channel at a glance.

That is the way it works currently, but based on the validation over at winget-pkgs, these would be completely separate package Identifiers. If I’m not mistaken, the whole point of channels is to combine package identifiers to be better representative of how packages are published, since some publishers do not allow side-by-side compatibility for different channels and so a user may not realize that Example.Package would be overwritten if they install Example.Package.Dev

Reply:
One that I could think of is Ubuntu. Canonical.Ubuntu `16.04`, `18.04`, and `20.04` all have their own release cadence, but so does the un-versioned channel. In this case, `20.04` would make the most sense to be the default channel.

Reply:
Is the thinking that the default channel will march forward with the latest release?  So currently it would be `20.04` but when `22.04` is released it would change to that, while still leaving the `20.04` channel as is.  I see the value there; if we used an empty channel then to achieve the same thing would require moving/renaming all of the manifests.

I think it still makes sense (and I think what you are saying the spec) to allow for the empty channel to be the default default.  Then *only* allow for `DefaultChannel` when there is no empty channel.